### PR TITLE
Product Image

### PR DESCRIPTION
**Pull request by gugabribeiro**
#### What is the purpose of this pull request?
Add the costumized ProductImage component that will display a list of product image thumbnails and a image that is selected according to the click on the thumbnails.

#### What problem is this solving?
Add the ProductImage component

#### How should this be manually tested?
```
$ npm t     # this
$ yarn test # or this one
```

#### Screenshots or example usage
The component workspace: https://product-image--storecomponents.myvtex.com/

**Vertical Thumbnail**

captura de tela de 2018-04-25 13-54-31

**Horizontal Thumbnail**

captura de tela de 2018-04-25 13-56-36

**Mobile**

captura de tela de 2018-04-25 13-57-01

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/ProductImage.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ *  Product Image component.
+ *  Display a list of thumbnail images in a slider and a main image of a product.
+ */
+class ProductImage extends Component {
+  constructor(props) {
+    super(props)
+    const images = this.props.images
+    
+    this.state = {
+      selectedImage: images[DEFAULT_SELECTED_IMAGE],
```
[the remark this answers is not in the document]
Comment (gugabribeiro) [reply react/ProductImage.js]: Eu conversei com Bruno sobre isso quando comecei a implementar o componente e ontem quando tu levantou a mesma questão no último pull request e ele afirmou que prefere que seja feito dessa forma estática mesmo, vai caber ao usuário fazer a ordenação das imagens do produto.

File: react/SelectedImage.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * Selected Image component.
+ * Display an image.
+ */
+class SelectedImage extends Component {
+  render() {
+    const { imageUrl, imageText } = this.props.image
+    const className = "dn di-ns w-80-ns"
+
+    return (
+      <div className={className}>
```
Comment (ClaudivanFilho) [inline react/SelectedImage.js]: coloca uma className aqui `vtex-product-image__selected-image`

File: react/ProductImage.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  render() {
+    const { images } = this.props
+
+    const className = this.configureClassName()
+    const selectedImage = this.configureSelectedImage()
+    const thumbnailSlider = this.configureThumbnailSlider()
+
+    return (
+      <div className={className}>
+        { images.length > 1 ? <ThumbnailSlider {...thumbnailSlider} /> : '' }
```
Comment (ClaudivanFilho) [inline react/ProductImage.js]: troca isso
 ``` { images.length > 1 ? <ThumbnailSlider {...thumbnailSlider} /> : '' } ```
por isso
 ``` { images.length && <ThumbnailSlider {...thumbnailSlider} /> } ```

File: react/ProductImage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ClaudivanFilho) [reply react/ProductImage.js]: Massa

File: react/ProductImage.js (hunk 3 above)
Comment (gugabribeiro) [reply react/ProductImage.js]: Mas nesse caso ele vai renderizar o thumbnail se houver 1 elemento na lista de imagens, o que não deve acontecer.

File: react/ThumbnailSlider.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+   */
+  configureSliderSettings = () => {
+    const { images, maxVisibleItems, orientation} = this.props
+  
+    const sliderVertical = (orientation === VERTICAL)
+
+    /** The number of visible slider items should not exceed the number of images, in the case that that condition is not satisfied, 
+     * then it must not exceed the maxVisibleItems property. Finnaly, the number of visible slider items should not be greater 
+     * than the maximum allowed value that is defined by MAX_VISIBLE_ITEMS */
+    const numOfVisivleItems = Math.min(maxVisibleItems, images.length, MAX_VISIBLE_ITEMS)
```
Comment (ClaudivanFilho) [inline react/ThumbnailSlider.js]: numOfVisivleItems -> numOfVisibleItems

File: react/ThumbnailSlider.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  
+  /** 
+   * Function that configure slider settings according to the component props
+   */
+  configureSliderSettings = () => {
+    const { images, maxVisibleItems, orientation} = this.props
+  
+    const sliderVertical = (orientation === VERTICAL)
+
+    /** The number of visible slider items should not exceed the number of images, in the case that that condition is not satisfied,
```
Comment (ClaudivanFilho) [inline react/ThumbnailSlider.js]: `that that`

File: react/ThumbnailSlider.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  /** 
+   * Function that configure slider settings according to the component props
+   */
+  configureSliderSettings = () => {
+    const { images, maxVisibleItems, orientation} = this.props
+  
+    const sliderVertical = (orientation === VERTICAL)
+
+    /** The number of visible slider items should not exceed the number of images, in the case that that condition is not satisfied, 
+     * then it must not exceed the maxVisibleItems property. Finnaly, the number of visible slider items should not be greater
```
Comment (ClaudivanFilho) [inline react/ThumbnailSlider.js]: finnaly -> finally

File: react/SelectedImage.js (hunk 2 above)
Comment (gugabribeiro) [reply react/SelectedImage.js]: Mantendo a className atual ou reescrevendo a className atual em ```vtex-product-image__selected-image```? Lembrando que algumas dessas classes possuem media queries do tachyons.

File: react/ProductImage.js (hunk 3 above)
Comment (ClaudivanFilho) [reply react/ProductImage.js]: faz assim então `{ images.length > 1 && <ThumbnailSlider {...thumbnailSlider} /> }`

File: react/package.json [hunk 7]
```diff
@@ -6,7 +6,9 @@
     "test": "jest"
   },
   "dependencies": {
-    "@vtex/styleguide": "2.0.0-rc.35"
+    "@vtex/styleguide": "2.0.0-rc.35",
```
Comment (ClaudivanFilho) [inline react/package.json]: coloca `"^2.0.0-rc.37"`

File: pages/pages.json [hunk 8]
```diff
@@ -1,14 +1,13 @@
 {
   "pages": {
-    "product-summary": {
+    "product-image": {
```
Comment (brunojdo) [inline pages/pages.json]: don't need this file.

File: react/ProductImage.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    { imageUrl: "https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro20.png", imageText: "" },
+    { imageUrl: "https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro21.png", imageText: "" },
+    { imageUrl: "https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro22.png", imageText: "" },
+    { imageUrl: "https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro23.png", imageText: "" },
+    { imageUrl: "https://raw.githubusercontent.com/vtex-apps/product-summary/feature/product-image/images/500x500-img-pro24.png", imageText: "" }
+  ],
+  thumbnailSliderOrientation: VERTICAL
+}
+
+ProductImage.schema = {
```
Comment (brunojdo) [inline react/ProductImage.js]: don't need this.

File: react/ThumbnailSlider.js (hunk 5 above)
Comment (brunojdo) [inline react/ThumbnailSlider.js]: "Number of...

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (brunojdo) [inline .vscode/settings.json]: .gitignore

File: react/package.json (hunk 7 above)
Comment (brunojdo) [reply react/package.json]: 39 😄

File: react/theme.css [hunk 11]
```diff
@@ -1,12 +1,67 @@
-@import url(//io.vtex.com.br/fonts/fabriga/stylesheet.css);
-body,button,input,optgroup,select,textarea,span,h1,h2,h3,h4,p1 {
-  font-family: Fabriga,-apple-system,BlinkMacSystemFont,avenir next,avenir,helvetica neue,helvetica,ubuntu,roboto,noto,segoe ui,arial,sans-serif
+@import url('~slick-carousel/slick/slick.css');
+@import url('~slick-carousel/slick/slick-theme.css');
+
+.vtex-product-image .vtex-product-image__selected-image
```
Comment (ClaudivanFilho) [inline react/theme.css]: pode deletar essas classes aqui, é so pra deixar definido no componente mesmo

File: react/__tests__/ProductImage.test.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      expect(
+        component.find('.slick-active').length
+      ).toBe(NUM_OF_VISIBLE_ITEMS)
+    })
+
+    it('should display the entire image of thumbnails if this number is less than the NUM_OF_VISIBLE_ITEMS that is passed as a props', () => {
+      const images = [ { imageUrl: "", imageText: "" },
+                       { imageUrl: "", imageText: "" } ]
+      const { component } = renderComponent({
+        images: images,
```
Comment (ClaudivanFilho) [inline react/__tests__/ProductImage.test.js]: `images, images` -> `images`

File: react/ProductImage.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+ProductImage.propTypes = {
+  /** Array of images to be passed for the Thumbnail Slider component as a props */
+  images: PropTypes.arrayOf(PropTypes.shape({
+    /** URL of the image */
+    imageUrl: PropTypes.string.isRequired,
+    /** Text that describes the image */
+    imageText: PropTypes.string.isRequired,
+  })),
```
Comment (ClaudivanFilho) [inline react/ProductImage.js]: coloca images como isRequired

File: react/ThumbnailItem.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    event.preventDefault()
+    const { image, onClick } = this.props
+    onClick(image)
+  }
+
+  render() {
+    const { imageUrl, imageText } = this.props.image
+
+    return (
+      <a href="#" onClick={this.handleClick}>
```
Comment (leticiawanderley) [inline react/ThumbnailItem.js]: Not sure about using an `a` tag if there's no link to be followed. Wouldn't a div work similarly?

File: react/Arrow.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+Arrow.propTypes = {
+  /** Css class of the element that is passed by the react-slick parent component. */
+  className: PropTypes.string,
+  /** Custom css class of the element. */
+  customClassName: PropTypes.string,
+  /** Custom style of the element. */
+  customStyle: PropTypes.object,
+  /** Maximum number of items in the shelf. */
+  onClick: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (gugabribeiro) [reply react/Arrow.js]: This onClick function is passed by Slider component that is provided by react-slick library. In some cases the Slider do not pass any function to the Arrow component, so, according to the Slider implementation it is not required.

File: react/ProductImage.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+      onThumbnailClick: this.handleThumbnailClick,
+      orientation: thumbnailSliderOrientation,
+      maxVisibleItems: thumbnailMaxVisibleItems,
+    }
+  }
+
+  render() {
+    const { images } = this.props
+
+    const className = this.configureClassName()
```
[the remark this answers is not in the document]
Comment (gugabribeiro) [reply react/ProductImage.js]: I was looking for some implementations here and found a lot of components that uses that approach, which made me think that this is some kind of pattern. But I agree with you.

File: react/ThumbnailItem.js (hunk 14 above)
Comment (gugabribeiro) [reply react/ThumbnailItem.js]: I was trying to use the on mouse hover behavior that the <a> tag have by default. But I realized that it is possible to use a <div> tag with some tachyons classes which provides it. 
Fixed!

File: react/Arrow.js [hunk 17]
```diff
@@ -0,0 +1,38 @@
+import React, { Component } from 'react'
```
Comment (ClaudivanFilho) [inline react/Arrow.js]: Usa o componente de @salesfelipe . Já tá disponível, se liga -> https://github.com/vtex-apps/carousel/blob/master/react/Carousel.js#L8

File: react/Dots.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from 'react'
```
Comment (ClaudivanFilho) [inline react/Dots.js]: Idem

**Pull request merged**